### PR TITLE
state Rest API

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

prepare for fine grained state access, especially 
1. table items
2. query path (state/resource/0x123/0x1::Collections::Collections/table/collection_name/token_id)

will add doc once API is complete. (modules, table items, query paths will follow)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

y

## Test Plan

unittests

### CODE REVIEW

File: api/src/tests/state_test.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{current_function_name, tests::new_test_context};
+
+#[tokio::test]
+async fn test_query_resource() {
```
Comment (sitalkedia) [inline api/src/tests/state_test.rs:7]: Thank you for adding these comprehensive tests.

**Review by sitalkedia — APPROVED** [review]

**Review by davidiw — CHANGES_REQUESTED** [review]
Overall looks good, could we please keep the API documentation up to date: https://github.com/aptos-labs/aptos-core/blob/main/api/doc/openapi.yaml

these things quickly diverge if we don't modify both in the same PR

File: storage/storage-interface/src/state_view.rs [hunk 2]
```diff
@@ -62,3 +62,24 @@ impl LatestDbStateView for Arc<dyn DbReader> {
         })
     }
 }
+
+pub trait DbStateViewAtVersion {
```
Comment (zekun000) [inline storage/storage-interface/src/state_view.rs:66]: why we need this?

**Comment by msmouse** [comment]
I don't want to open it yet.. might change from GET to POST (to take in complex table keys like the `ID`, and long query paths), are we okay for these breaking changes in the sdks?

**Comment by davidiw** [comment]
> I don't want to open it yet.. might change from GET to POST (to take in complex table keys like the `ID`, and long query paths), are we okay for these breaking changes in the sdks?

I'm a bit concerned about moving too fast here since there's an ever growing set of ecosystem tools It might be good for us to quickly align on the scope and nature here as well as a strategy, ideally we can do as much plumbing behind the doors and then help the community quickly transfer over to the new paradigm.

File: storage/storage-interface/src/state_view.rs (hunk 2 above)
Comment (msmouse) [reply storage/storage-interface/src/state_view.rs:66]: When we do multi-hop queries (like indirection to a table item), we want to base the queries on a consistent db view.

The current implementation under /account is wrong -- although it doesn't do any indirection, it always interprets resources at a certain version with the LATEST modules. (Before the latest `latest_state_view()` thing was introduced, it was worse -- the converter fetches different modules at changing versions.)

**Comment by msmouse** [comment]
Thought this without generating the APIs in the sdks is "behind the door". But anyway we can just hold on landing this and build in the PR everything.

**Comment by msmouse** [comment]
..... sorry, it landed.. we shouuld've issued '/cancel'

**Pull request merged**